polkadot/api/src/full.rs

```diff
@@ -18,46 +18,61 @@
 
 use client::backend::LocalBackend;
 use client::block_builder::BlockBuilder as ClientBlockBuilder;
-use client::{Client, LocalCallExecutor};
+use client::{self, Client, LocalCallExecutor, CallExecutor};
+use codec::{Encode, Decode};
 use polkadot_executor::Executor as LocalDispatch;
 use substrate_executor::NativeExecutor;
-use state_machine;
+use state_machine::ExecutionManager;
 
 use runtime::Address;
-use runtime_primitives::traits::AuxLookup;
 use primitives::{
 	AccountId, Block, Header, BlockId, Hash, Index, InherentData,
 	SessionKey, Timestamp, UncheckedExtrinsic,
 };
 use primitives::parachain::{DutyRoster, Id as ParaId};
-
-use {BlockBuilder, PolkadotApi, LocalPolkadotApi, ErrorKind, Error, Result};
-
-// set up the necessary scaffolding to execute a set of calls to the runtime.
-// this creates a new block on top of the given ID and initialises it.
-macro_rules! with_runtime {
-	($client: ident, $at: expr, $exec: expr) => {{
-		let parent = $at;
-		let header = Header {
-			parent_hash: $client.block_hash_from_id(&parent)?
-				.ok_or_else(|| ErrorKind::UnknownBlock(format!("{:?}", parent)))?,
-			number: $client.block_number_from_id(&parent)?
+use {BlockBuilder, PolkadotApi, LocalPolkadotApi, Error, ErrorKind, Result};
+
+fn call<B, R>(client: &Client<B, LocalCallExecutor<B, NativeExecutor<LocalDispatch>>, Block>, at: &BlockId, method: &'static str, input: &[u8]) -> Result<R>
```

Comment:
need an issue to move this logic, and much of the API, into substrate.

Reply:
this trait is meant to be wrappers for the Polkadot-runtime specific functions. if there are any functions here that are general to substrate, they should indeed be moved there.

Reply:
ok - just the bits of the api that are substrate-specific. but the logic for calling into the api, nonetheless should be moved to substrate.

Reply:
all the block authorship APIs are already used from substrate. the `PolkadotApi` has a function for giving you a `substrate_client::BlockBuilder` with inherent extrinsics already pushed. I'm not sure if the `inherent_extrinsics` bit can be extracted because it's reliant on polkadot-specific data. but we could make it part of a more general substrate authorship API suite.

Reply:
the calling logic should be extracted out, but moving those APIs exclusively to substrate could diminish the value of this crate, which is to abstract a bit over substrate for testing. I would rather have this trait wrap substrate functionality for calling into runtimes. This trait currently provides all functions necessary for block authorship. If some functionality was moved out of this trait, it would complicate the block authorship code by adding additional trait bounds on this generic parameter.

Reply:
> the calling logic should be extracted out

cool.